internal/cephfs/snapshot.go

```diff
@@ -0,0 +1,272 @@
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix + cr.ID,
+		"--keyfile=" + cr.KeyFile,
+		"--force",
+	}
+
+	err := execCommandErr(
+		ctx,
+		"ceph",
+		args[:]...)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete subvolume snapshot %s %s(%s) in fs %s"), string(snapID), string(volID), err, volOptions.FsName)
+		return err
+	}
+	return nil
+}
+
+type snapshotInfo struct {
+	CreatedAt        string `json:"created_at"`
+	CreationTime     *timestamp.Timestamp
+	DataPool         string `json:"data_pool"`
+	HasPendingClones string `json:"has_pending_clones"`
+	Protected        string `json:"protected"`
+	Size             int    `json:"size"`
+}
+
+func getSnapshotInfo(ctx context.Context, volOptions *volumeOptions, cr *util.Credentials, snapID, volID volumeID) (snapshotInfo, error) {
+	snap := snapshotInfo{}
+	args := []string{
+		"fs",
+		"subvolume",
+		"snapshot",
+		"info",
+		volOptions.FsName,
+		string(volID),
+		string(snapID),
+		"--group_name",
+		volOptions.SubvolumeGroup,
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix + cr.ID,
+		"--keyfile=" + cr.KeyFile,
+		"--format=json",
+	}
+	err := execCommandJSON(
+		ctx,
+		&snap,
+		"ceph",
+		args[:]...)
+	if err != nil {
+		if strings.Contains(err.Error(), ErrSnapNotFound.Error()) {
+			return snapshotInfo{}, err
+		}
+		klog.Errorf(util.Log(ctx, "failed to get subvolume snapshot info %s %s(%s) in fs %s"), string(snapID), string(volID), err, volOptions.FsName)
+		return snapshotInfo{}, err
+	}
+	return snap, nil
+}
+
+func protectSnapshot(ctx context.Context, volOptions *volumeOptions, cr *util.Credentials, snapID, volID volumeID) error {
+	args := []string{
+		"fs",
+		"subvolume",
+		"snapshot",
+		"protect",
+		volOptions.FsName,
+		string(volID),
+		string(snapID),
+		"--group_name",
+		volOptions.SubvolumeGroup,
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix + cr.ID,
+		"--keyfile=" + cr.KeyFile,
+	}
+
+	err := execCommandErr(
+		ctx,
+		"ceph",
+		args[:]...)
+	if err != nil {
+		if strings.Contains(err.Error(), ErrSnapProtectionExist.Error()) {
+			return nil
+		}
+		klog.Errorf(util.Log(ctx, "failed to protect subvolume snapshot %s %s(%s) in fs %s"), string(snapID), string(volID), err, volOptions.FsName)
+		return err
+	}
+	return nil
+}
+
+func unprotectSnapshot(ctx context.Context, volOptions *volumeOptions, cr *util.Credentials, snapID, volID volumeID) error {
+	args := []string{
+		"fs",
+		"subvolume",
+		"snapshot",
+		"unprotect",
+		volOptions.FsName,
+		string(volID),
+		string(snapID),
+		"--group_name",
+		volOptions.SubvolumeGroup,
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix + cr.ID,
+		"--keyfile=" + cr.KeyFile,
+	}
+
+	err := execCommandErr(
+		ctx,
+		"ceph",
+		args[:]...)
+	if err != nil {
+		if strings.Contains(err.Error(), ErrSnapProtectionExist.Error()) {
```

Comment:
This check seems to be wrong. why we are checking snap protection exist in unprotect?

Reply:
@Madhu-1 if snap is already unprotected ErrSnapProtectionExist is the error code. so we are safe and we dont need retrun an error. Rather go ahead with deletion. Let me document it..https://tracker.ceph.com/issues/46733